File: apps/android/app/src/main/java/org/llamenos/hotline/telephony/LinphoneService.kt
package org.llamenos.hotline.telephony

import android.content.Context
import android.util.Log
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import org.linphone.core.Call
import org.linphone.core.Core
import org.linphone.core.CoreListenerStub
import org.linphone.core.Factory
import org.linphone.core.MediaEncryption
import org.llamenos.hotline.di.ApplicationScope
import org.llamenos.hotline.hub.ActiveHubState
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton

data class SipTokenResponse(
    val username: String,
    val domain: String,
    val password: String,
    val transport: String,
    val expiry: Int,
)

@Singleton
class LinphoneService @Inject constructor(
    @ApplicationContext private val context: Context,
    private val activeHubState: ActiveHubState,
    @ApplicationScope private val scope: CoroutineScope,
) {
    private var core: Core? = null
    private val hubAccounts = ConcurrentHashMap<String, org.linphone.core.Account>()
    // TODO: Add eviction (e.g. LruCache) for entries that are never consumed
    // (call push arrives but Linphone never connects). Low risk for now given
    // low per-session call volume on a crisis line.
    private val pendingCallHubIds = ConcurrentHashMap<String, String>()  // callId → hubId

    fun initialize() {
        try {
            val factory = Factory.instance()
            val core = factory.createCore(null, null, context)
            core.isCallkitEnabled = true
            core.mediaEncryption = MediaEncryption.SRTP
            core.isMediaEncryptionMandatory = true

            core.audioPayloadTypes.forEach { pt ->
                pt.enable(pt.mimeType == "opus" || pt.mimeType == "PCMU")
            }

            setupCoreListener(core)
            core.start()
            this.core = core
            Log.i("LinphoneService", "Linphone Core initialized successfully")
        } catch (e: Exception) {
            Log.e("LinphoneService", "Failed to initialize Linphone Core: ${e.message}")
        }
    }

    fun registerHubAccount(hubId: String, sipParams: SipTokenResponse) {
        val core = this.core ?: run {
            Log.w("LinphoneService", "Core not initialized — cannot register SIP account for $hubId")
            return
        }
        try {
            val params = core.createAccountParams()
            val identity = Factory.instance().createAddress(
                "sip:${sipParams.username}@${sipParams.domain}"
            )
            params.identityAddress = identity
            val server = Factory.instance().createAddress(
                "sip:${sipParams.domain};transport=${sipParams.transport}"
            )
            params.serverAddress = server
            params.isRegisterEnabled = true
            val account = core.createAccount(params)
            core.addAccount(account)
            hubAccounts[hubId] = account
            Log.i("LinphoneService", "Registered SIP account for hub $hubId")
        } catch (e: Exception) {
            Log.e("LinphoneService", "Failed to register SIP account for $hubId: ${e.message}")
        }
    }

    fun unregisterHubAccount(hubId: String) {
        val account = hubAccounts.remove(hubId) ?: return
        core?.removeAccount(account)
        Log.i("LinphoneService", "Unregistered SIP account for hub $hubId")
    }

    fun storePendingCallHub(callId: String, hubId: String) {
        pendingCallHubIds[callId] = hubId
    }

    private fun setupCoreListener(core: Core) {
        core.addListener(object : CoreListenerStub() {
            override fun onCallStateChanged(
                core: Core,
                call: Call,
                state: Call.State,
                message: String,
            ) {
                val callId = call.callLog?.callId ?: return
                when (state) {
                    Call.State.IncomingReceived -> {
                        pendingCallHubIds.remove(callId)?.let { hubId ->
                            scope.launch { activeHubState.setActiveHub(hubId) }
                        }
                    }
                    Call.State.Released, Call.State.End -> {
                        pendingCallHubIds.remove(callId)
                    }
                    else -> {}
                }
            }
        })
    }

    internal fun pendingCallHubIdForTesting(callId: String): String? = pendingCallHubIds[callId]
    internal fun consumePendingCallHubForTesting(callId: String) { pendingCallHubIds.remove(callId) }
}

File: apps/ios/Sources/Services/LinphoneService.swift
import Foundation

#if canImport(linphonesw)
import linphonesw
#endif

// MARK: - LinphoneServiceProtocol

/// Protocol for SIP account lifecycle operations. Implemented by `LinphoneService` for
/// production and by test doubles in unit tests.
protocol LinphoneServiceProtocol: AnyObject {
    func registerHubAccount(hubId: String, sipParams: SipTokenResponse) throws
    func unregisterHubAccount(hubId: String)
    func handleVoipPush(callId: String, hubId: String)
}

// MARK: - LinphoneError

enum LinphoneError: LocalizedError {
    case notInitialized
    case accountRegistrationFailed(String)
    case coreStartFailed(String)

    var errorDescription: String? {
        switch self {
        case .notInitialized:
            return "Linphone Core not initialized"
        case .accountRegistrationFailed(let msg):
            return "SIP account registration failed: \(msg)"
        case .coreStartFailed(let msg):
            return "Linphone Core failed to start: \(msg)"
        }
    }
}

// MARK: - SipTokenResponse

/// SIP credentials returned by `GET /api/hubs/{hubId}/telephony/sip-token`.
/// Used to register the volunteer's SIP account for the active shift.
struct SipTokenResponse: Decodable {
    let username: String
    let domain: String
    let password: String
    let transport: String
    let expiry: Int
}

// MARK: - LinphoneService

/// Manages a Linphone SIP core instance for real-time VoIP call handling.
///
/// One SIP account is registered per hub the volunteer is on-shift for.
/// The `pendingCallHubIds` map correlates incoming VoIP push payloads
/// (which carry a `callId` and `hubId`) with the Linphone `Call` object
/// that subsequently arrives in `onCallStateChanged`, so the hub context
/// can be switched before the call is presented to the volunteer.
///
/// Conditional compilation (`#if canImport(linphonesw)`) ensures the file
/// compiles without the Linphone XCFramework. Run `scripts/download-linphone-ios.sh`
/// and re-run `xcodegen generate` to link the real SDK.
@Observable
final class LinphoneService: LinphoneServiceProtocol {
    // MARK: - Private State

    /// callId → hubId map set by VoIP push before Linphone fires onCallStateChanged.
    private var pendingCallHubIds: [String: String] = [:]
    private let pendingCallLock = NSLock()
    private weak var hubContext: HubContext?

    #if canImport(linphonesw)
    private var core: Core?
    private var hubAccounts: [String: Account] = [:]
    #endif

    // MARK: - Initialization

    init() {}

    /// Configure the service with an app-wide HubContext and start the Linphone Core.
    /// Call this once from AppState or LlamenosApp after the hub context is ready.
    func initialize(hubContext: HubContext) throws {
        self.hubContext = hubContext
        #if canImport(linphonesw)
        let factory = Factory.Instance
        let core = try factory.createCore(
            configFilename: "linphone",
            factoryConfigFilename: nil,
            systemContext: nil
        )
        core.callKitEnabled = true
        core.mediaEncryption = .SRTP
        core.mediaEncryptionMandatory = true

        // Allow only Opus and G.711 µ-law; disable all others.
        for pt in core.audioPayloadTypes {
            pt.enable(pt.mimeType == "opus" || pt.mimeType == "PCMU")
        }

        setupCoreDelegate(core: core)
        try core.start()
        self.core = core
        #endif
    }

    // MARK: - SIP Account Management

    /// Register a SIP account for the given hub. Called when the volunteer clocks in.
    func registerHubAccount(hubId: String, sipParams: SipTokenResponse) throws {
        #if canImport(linphonesw)
        guard let core else { throw LinphoneError.notInitialized }
        let params = try core.createAccountParams()
        let identity = try Factory.Instance.createAddress(
            addr: "sip:\(sipParams.username)@\(sipParams.domain)"
        )
        try params.setIdentityaddress(newValue: identity)
        let server = try Factory.Instance.createAddress(
            addr: "sip:\(sipParams.domain);transport=\(sipParams.transport)"
        )
        try params.setServeraddress(newValue: server)
        params.registerEnabled = true
        let account = try core.createAccount(params: params)
        try core.addAccount(account: account)
        hubAccounts[hubId] = account
        #endif
    }

    /// Unregister the SIP account for the given hub. Called when the volunteer clocks out.
    func unregisterHubAccount(hubId: String) {
        #if canImport(linphonesw)
        guard let account = hubAccounts.removeValue(forKey: hubId) else { return }
        core?.removeAccount(account: account)
        #endif
    }

    // MARK: - VoIP Push Handling

    /// Record the hub ID that triggered a VoIP push for a given call ID.
    /// Called by PushKit before Linphone fires `onCallStateChanged(.IncomingReceived)`.
    func handleVoipPush(callId: String, hubId: String) {
        pendingCallLock.lock()
        defer { pendingCallLock.unlock() }
        pendingCallHubIds[callId] = hubId
    }

    // MARK: - Core Delegate (Linphone)

    #if canImport(linphonesw)
    private func setupCoreDelegate(core: Core) {
        let delegate = CoreDelegateStub(
            onCallStateChanged: { [weak self] _, call, state, _ in
                guard let self else { return }
                let callId = call.callLog?.callId ?? ""
                switch state {
                case .IncomingReceived:
                    self.pendingCallLock.lock()
                    let hubId = self.pendingCallHubIds.removeValue(forKey: callId)
                    self.pendingCallLock.unlock()
                    if let hubId {
                        Task { @MainActor in
                            self.hubContext?.setActiveHub(hubId)
                        }
                    }
                case .Released, .End:
                    self.pendingCallLock.lock()
                    self.pendingCallHubIds.removeValue(forKey: callId)
                    self.pendingCallLock.unlock()
                default:
                    break
                }
            }
        )
        core.addDelegate(delegate: delegate)
    }
    #endif

    // MARK: - Test-only Accessors

    #if DEBUG
    /// Returns the pending hub ID for a call ID without consuming it. For unit tests only.
    func pendingCallHubIdForTesting(_ callId: String) -> String? {
        pendingCallLock.lock()
        defer { pendingCallLock.unlock() }
        return pendingCallHubIds[callId]
    }

    /// Remove the pending hub ID for a call ID, simulating post-consumption. For unit tests only.
    func consumePendingCallHubForTesting(_ callId: String) {
        pendingCallLock.lock()
        pendingCallHubIds.removeValue(forKey: callId)
        pendingCallLock.unlock()
    }
    #endif
}
